File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def find(x,y):
            if parent[(x,y)] != (x,y):
                parent[(x,y)] = find(parent[(x,y)][0], parent[(x,y)][1])
                
            return parent[(x,y)]
        
        def union(x,y,x_,y_):
            setX = find(x,y)
            setY = find(x_,y_)
            
            if setX != setY:
                if size[setX] >= size[setY]:
                    size[setX]+=size[setY]
                    parent[setY] = setX
                    
                else:
                    size[setY]+=size[setX]
                    parent[setX] = setY
                
        
        M,N = len(grid), len(grid[0])
        parent = {(x,y):(x,y) for x in range(M) for y in range(N) if grid[x][y] == "1"}
        size = {(x,y):1 for x in range(M) for y in range(N) if grid[x][y] == "1"}
        
        dir = [(-1,0),(1,0),(0,1),(0,-1)]
        
        for i in range(M):
            for j in range(N):
                if grid[i][j] == "1":
                    for dx,dy in dir:
                        dx+=i 
                        dy+=j
                        
                        if 0<=dx<M and 0<=dy<N and grid[dx][dy] == "1":
                            union(i,j,dx,dy)
        
        return (sum([1 for k,v in parent.items() if k==v]))

File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    vector<pair<int,int>> dir = {{-1,0},{1,0},{0,-1},{0,1}};
    
    void dfs(vector<vector<char>> &grid, int i, int j, int m, int n){
        if(i<0 or i>=m or j<0 or j>=n or grid[i][j]=='0')
            return;
        
        grid[i][j] = '0';
        
        for(auto &z:dir){
            int x = z.first+i;
            int y = z.second+j;
            
            dfs(grid,x,y,m,n);
        }
    }
    
    void bfs(vector<vector<char>> &grid, int i, int j, int m, int n){
        queue<pair<int,int>> q;
        
        q.push({i, j});
        grid[i][j] = '0';
        
        while(!q.empty()){
            int size = q.size();
            
            for(int i=0;i<size;i++){
                pair<int,int> loc = q.front();
                q.pop();
                
                for(auto &z:dir){
                    int x = z.first+loc.first;
                    int y = z.second+loc.second;
                    
                    if(x<0 or x>=m or y<0 or y>=n or grid[x][y]=='0')
                        continue;
                    
                    grid[x][y] = '0';
                    
                    q.push({x,y});
                }
            }
        }
    }

    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size(), n = grid[0].size();
        int cnt = 0;
        
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j]=='1'){
                    cout<<i<<" "<<j<<endl;
                    bfs(grid,i,j,m,n);
                    cnt+=1;
                }
            }
        }
        return cnt;
    }
};